include/cpp2util.h

```diff
@@ -1736,20 +1739,7 @@ inline constexpr auto is( X const& x, bool (*value)(X const&) ) -> bool {
 //  If it's confusing, we can switch this to <From, To>
 
 template< typename To, typename From >
-inline constexpr auto is_narrowing_v =
-    // [dcl.init.list] 7.1
-    (std::is_floating_point_v<From> && std::is_integral_v<To>) ||
-    // [dcl.init.list] 7.2
-    (std::is_floating_point_v<From> && std::is_floating_point_v<To> && sizeof(From) > sizeof(To)) ||
-    // [dcl.init.list] 7.3
-    (std::is_integral_v<From> && std::is_floating_point_v<To>) ||
-    (std::is_enum_v<From> && std::is_floating_point_v<To>) ||
-    // [dcl.init.list] 7.4
-    (std::is_integral_v<From> && std::is_integral_v<To> && sizeof(From) > sizeof(To)) ||
-    (std::is_enum_v<From> && std::is_integral_v<To> && sizeof(From) > sizeof(To)) ||
-    // [dcl.init.list] 7.5
-    (std::is_pointer_v<From> && std::is_same_v<To, bool>)
```

Comment:
Doesn't this lose the pointer to bool test?  What about the enum handling?

Reply:
Hm... yes, this is lost. Interestingly, in our tests and in cppfront, there is no such call to cast ptr to bool with as().
If needed, I can bring it back.

Reply:
It would be good to have tests for things which we know should fail.

Reply:
`as bool` seems nicer than `!= nullptr`, although the latter seems more explicit.

Reply:
On the other hand, it is perceived by the compiler as `narrowing`, so maybe it should require special treatment with `unsafe_narrow<bool>(ptr)`
```
<source>:59:19: warning: narrowing conversion of '& i' from 'int*' to 'bool' [-Wnarrowing]
   59 |     auto b = bool{&i};
```

Reply:
I might have misunderstood you. In my current implementation, you cannot cast a pointer to bool. It ends up with an error:
```
cppfront/include/cpp2util.h:3515:9: error: static assertion failed due to requirement 'program_violates_type_safety_guarantee<bool, int *>': No safe 'as' cast available - please check your cast
        static_assert(
        ^
narrowing.cpp2:15:31: note: in instantiation of function template specialization 'cpp2::impl::as_<bool, int *>' requested here
    std::cout << (cpp2::impl::as_<bool>(cpp2::move(pi))) << std::endl;
```

The previous implementation also considers it as narrowing, so there is no change in behavior.

Check here: https://godbolt.org/z/5KzMPevrE

Reply:
As long as `is_narrowing_v` is now only ever used for arithmetic types, then you're good with what you have, but then you don't really need `arithmetic<To> && arithmetic<From> &&` and `!brace_initializable_to<From,To>` is sufficient, and in that case, do you even need `is_narrowing_v`?  You can just use `!brace_initializable_to`.

I was assuming that `is_narrowing_v` was used for things other than arithmetic due to all the checks that were there before.

Reply:
Now that I've gone looking at where this is used, I think the current behavior is actually correct.  Conversion from `int` to `size_t` is a runtime check:

```
    //  Signed/unsigned conversions to a not-smaller type are handled as a precondition,
    //  and trying to cast from a value that is in the half of the value space that isn't
    //  representable in the target type C is flagged as a type_safety contract violation
    else if constexpr (
        std::is_integral_v<C> &&
        std::is_integral_v<CPP2_TYPEOF(x)> &&
        std::is_signed_v<CPP2_TYPEOF(x)> != std::is_signed_v<C> &&
        sizeof(CPP2_TYPEOF(x)) <= sizeof(C)
    )
    {
        const C c = static_cast<C>(CPP2_FORWARD(x));
        type_safety.enforce(   // precondition check: must be round-trippable => not lossy
            static_cast<CPP2_TYPEOF(x)>(c) == x && (c < C{}) == (x < CPP2_TYPEOF(x){}),
            "dynamic lossy narrowing conversion attempt detected" CPP2_SOURCE_LOCATION_ARG_AS
        );
        return CPP2_COPY(c);
    }
```

Reply:
Yes, you are right. I try to minimize the change to make it easier for @hsutter to merge (I have prepared big PRs in the past that didn't go through). The current `is_narrowing_v` is only for arithmetic types. If you remove it, then the `as_()` function will also match for non-arithmetic types that will need additional treatment unrelated to arithmetic types. 
```cpp
template< typename C >
inline constexpr auto as_( auto&& x ) -> decltype(auto)
{
    if constexpr (is_narrowing_v<C, CPP2_TYPEOF(x)>) {
        static_assert(
            program_violates_type_safety_guarantee<C, CPP2_TYPEOF(x)>,
            "'as' does not allow unsafe possibly-lossy narrowing conversions - if you're sure you want this, use 'unsafe_narrow<T>' to explicitly force the conversion and possibly lose information"
        );
    }
    else if constexpr (is_unsafe_pointer_conversion_v<C, CPP2_TYPEOF(x)>)
//...
```

I will refactor more `is()` and `as()` so I will get there soon.

I have my big PR on my local branch, which I am introducing here step by step :) (adjusting it to @hsutter vision).

Reply:
> Now that I've gone looking at where this is used, I think the current behavior is actually correct. Conversion from `int` to `size_t` is a runtime check:

No, it is not correct. When you use `as` cast, it should fail as soon as possible. Just because `int` is signed and `size_t` is unsigned, the conversion might fail. We do similar checks for ints of various sizes. E.g. you cannot cast std::int16_t` to `std::int8_t`.

Check the following example:
```cpp
i: std::int16_t = 12;
j := i as std::int8_t;
```
it ends with:
```cpp
cppfront/include/cpp2util.h:3515:9: error: static assertion failed due to requirement 'program_violates_type_safety_guarantee<signed char, short>': 'as' does not allow unsafe possibly-lossy narrowing conversions - if you're sure you want this, use 'unsafe_narrow<T>' to explicitly force the conversion and possibly lose information
        static_assert(
        ^
narrowing.cpp2:18:25: note: in instantiation of function template specialization 'cpp2::impl::as_<signed char, short>' requested here
    auto j {cpp2::impl::as_<std::int8_t>(cpp2::move(i))}; 
                        ^
1 error generated.
```

Reply:
It works differently when you use numeric literal. The following code is correct:
```cpp
s := 123 as size_t;
```
We can check if `123` literal fits in `size_t` at compile-time, so it works.
and because of that the following code:
```cpp
s := -123 as size_t;
```
will fail on compile time with message:
```
cppfront/include/cpp2util.h:3529:13: error: static assertion failed due to requirement 'program_violates_type_safety_guarantee<unsigned long, int>': 'as' does not allow unsafe possibly-lossy narrowing conversions - if you're sure you want this, use `unsafe_narrow<T>()` to explicitly force the conversion and possibly lose information
            static_assert(
            ^
narrowing.cpp2:19:25: note: in instantiation of function template specialization 'cpp2::impl::as_<unsigned long, -123>' requested here
    auto s {cpp2::impl::as_<size_t, -123>()};
```

Reply:
> We do similar checks for ints of various sizes. E.g. you cannot cast std::int16_t to std::int8_t.

Yes, larger to smaller is compile time, but according to the code I quoted, when casting between different signed-ness and not going larger to smaller, it's runtime.

`int` to `size_t` should hit this code:

https://github.com/hsutter/cppfront/blob/main/include/cpp2util.h#L1834-L1871

Oh, there's another `as_` later which uses this.  What's the difference between `as` and `as_`?  I can't find any uses for `as_`.

Reply:
Static assert should be triggered whenever "from" type can represents values that cannot be represented by "to" type. That is why u16 to u8 will trigger assertion (it could be also runtime check but we decided to trigger it as soon as possible - so compile time). 

`as_` is generated by cppfront where you use `as` (check generated cpp file). It is old issue: https://github.com/hsutter/cppfront/pull/106#issuecomment-1373962682

Reply:
> as_ is generated by cppfront where you use as (check generated cpp file).  I tried

Reply:
>Static assert should be triggered whenever "from" type can represents values that cannot be represented by "to" type. That is why u16 to u8 will trigger assertion (it could be also runtime check but we decided to trigger it as soon as possible - so compile time).

That doesn't explain the explicit runtime check in the code for "a type with different unsignedness but same or larger size".  Is that code not actually being used?